### ISSUE TITLE
Some General Questions

### ISSUE THREAD

**Issue by Glaringsoul** [KohakuBlueleaf/LyCORIS#23]
This isn't so much an Issue, but rather starting a thread in here;

First of all, Great work! 
After reading through the Algo.md & Demo.md it looks really promising.

But i still have some Questions left open:

### **_What *exactly* is LyCORIS?:_**

Because as i Understood it:

1. LoRA: The "Normal" Baseline,
2. LoCon: Introduction of a Convolution Network, which allows the model to also train the Res-Blocked parts, which "prevent" overspecialisations and in theory should also increase the Flexibility. (Depracted and turned into *This*)
3. LoHa: Utilizing the Hadamard Product representation, which allows for better weighting generation, and also saves Memory
4. CP-Decomposition: Speeding up the CPU, with negligible Accuracy loss
5. LyCORIS: Convolution Network (LoCon) + Hadamard Product Representation (LoHa) + CP-decomposition

Which together means:

1. Better Learning Rates
2. More Flexible final Product
3. Less Memory used
4. Faster Processing

Or am i Missing/ Misunderstanding Something?

------------------------------------------------------------------------

As it Currently stands LyCORIS only works for SD; Does that Include SD-Based models like Anything or HD, or only the "Real" Stable Diffusion Releases?

~~Also less of a question, but how will your project be maintained in the future, as i didn't see any update function?~~

~~And will you Co-Operate with Bmaltais again, to make Lycoris a Standard Feature of the GUI, just like LoCon?~~
Edit: OK i'm stupid After updating Bmaltais GUI the option is now there, without the need to specifically install from the directory, as the Gui handles it itself including (I assume) updating itself...

Sorry if some of those questions sound stupid, but i'm neither a Mathematician, nor a Computer Engineer, and i have a fairly hard time wrapping my head around a lot of the inner workings of what exactly is happening.

**Comment by kaibioinfo** [comment]
Cool, I had exactly the same questions ;D

Also: If I understand it correctly, Hadamard allows for higher rank approximation with less parameters. However, in the original LORA paper they stated that for finetuning on special tasks, the rank is extremely low anyways and more parameters usually just end up in more noise. Do you have any experiences that using Hadamard is leading to better models or, the other possibility, to more overfitting?

**Comment by Glaringsoul** [comment]
Me personally no,
I dabbled a bit around with creating style LoRA's, just stumbled across this.

Considering Hadamard (according to the paper) as an "Upgrade", the result should be more compact and also more flexible, due to it having a larger capacity:

> In this work, we propose a communication-efficient parameterization, FedPara, for federated learning (FL) to overcome the burdens on frequent model uploads and downloads. Our method re-parameterizes weight parameters of layers using low-rank weights followed by the Hadamard product. Compared to the conventional low-rank parameterization, our FedPara method is not restricted to low-rank constraints, and thereby it has a far larger capacity. This property enables to achieve comparable performance while requiring 3 to 10 times lower communication costs than the model with the original layers, which is not achievable by the traditional low-rank methods. The efficiency of our method can be further improved by combining with other efficient FL optimizers. In addition, we extend our method to a personalized FL application, pFedPara, which separates parameters into global and local ones. We show that pFedPara outperforms competing personalized FL methods with more than three times fewer parameters. (From [https://arxiv.org/abs/2108.06098](https://arxiv.org/abs/2108.06098))

The Idea that this could lead to more Overfitting is interesting though, but should be fairly manageable by utilizing a lower Learning rate (what you should normally do for LoHa anyway), and just Lowering the Epoch/Step Count.

Also overfitting should (in theory) also be Counteracted by the Convolution network re-introducing Res-Blocked parts.

Not fully sure if that is how it works, but in theory overfitting is a symptom of too much training, and if LyCORIS cuts down on the training time by automatically making training more efficient i'm all for it.

I'm gonna assume we will have to find new training Parameters that "Work well", but that is a small price to pay for the Upsides we should be getting...

**Comment by KohakuBlueleaf** [comment]
1, True
2, Basically yes. just train lora on conv.
3, True, False. Hadamard product representation need a little bit more memory.
4, cp-decomposition is another way to decompose the multi-dim tensor (like the weight for convolution layer). Allows you to get same rank with more less parameter, and take less computation power.
5, LyCORIS is the name for this project, I will continuously add new algo into this project.

Can this answer your question?

**Comment by kaibioinfo** [comment]
Another question, I don't want to open a new issue as I'm not sure if its really a bug, but:
```python

    def forward(self, x):
        if self.cp:
            return self.dropout(
                self.org_forward(x) 
                + self.lora_up(self.lora_mid(self.lora_down(x)))
            ) * self.multiplier * self.scale
        else:
            bias = None if self.org_module[0].bias is None else self.org_module[0].bias.data
            return self.op(
                x,
                (self.org_module[0].weight.data 
                + self.dropout(self.make_weight()) * self.multiplier * self.scale),
                bias,
                **self.extra_args,
            )
```
why is in the cp case the output of lora+original output multiplied with the multiplier? Shouldn't it be
```python
return self.dropout(
                self.org_forward(x) 
                + self.lora_up(self.lora_mid(self.lora_down(x))) * self.multiplier * self.scale
            )
```

Sorry, if this is a stupid question.

**Comment by Glaringsoul** [comment]
> 

You should maybe mention *where* its from,
Namely Locon.py line 86f

The Locon.py out of the archive indeed had
`
def forward(self, x):
        bias = None if self.org_module[0].bias is None else self.org_module[0].bias.data
        return self.op(
            x,
            (self.org_module[0].weight.data 
             + self.dropout(self.make_weight()) * self.multiplier * self.scale),
            bias,
            **self.extra_args,
        )
`
which means you are correct?

I'm going to leave the honors of creating a Pull Request and Issue to you.

Nice Catch !

**Comment by Glaringsoul** [comment]
> Another question, I don't want to open a new issue as I'm not sure if its really a bug, but:
> 
> ```python
>     def forward(self, x):
>         if self.cp:
>             return self.dropout(
>                 self.org_forward(x) 
>                 + self.lora_up(self.lora_mid(self.lora_down(x)))
>             ) * self.multiplier * self.scale
>         else:
>             bias = None if self.org_module[0].bias is None else self.org_module[0].bias.data
>             return self.op(
>                 x,
>                 (self.org_module[0].weight.data 
>                 + self.dropout(self.make_weight()) * self.multiplier * self.scale),
>                 bias,
>                 **self.extra_args,
>             )
> ```
> 
> why is in the cp case the output of lora+original output multiplied with the multiplier? Shouldn't it be
> 
> ```python
> return self.dropout(
>                 self.org_forward(x) 
>                 + self.lora_up(self.lora_mid(self.lora_down(x))) * self.multiplier * self.scale
>             )
> ```
> 
> Sorry, if this is a stupid question.

I think you can go ahead and crate an Issue/ Pull Request for that.

I have no idea if that is actually correct or not, but i highly assume it is.

I would only have mentioned where to find it, namely:

`venv\lib\site-packages\lycoris\locon.py`
in Line 86 following.

Band aid fix until a new version that is more than a single bugfix is read to be release, would be to manually edit the two lines to fix it...

**Comment by kaibioinfo** [comment]
yeah, after reading a paper about cp-decomposition on convolutional neural networks I'm pretty sure the fix is correct. I made a pull request.
